rust/lance/src/session/index_extension.rs

```diff
@@ -5,12 +5,34 @@ use std::sync::Arc;
 
 use lance_core::Result;
 use lance_file::reader::FileReader;
+use lance_index::{IndexParams, IndexType};
 
 use crate::{index::vector::VectorIndex, Dataset};
 
+pub trait IndexExtension: Send + Sync {
+    fn index_type(&self) -> IndexType;
+
+    fn to_scalar_index_extension(self: Arc<Self>) -> Option<Arc<dyn ScalarIndexExtension>>;
+
+    fn to_vector_index_extension(self: Arc<Self>) -> Option<Arc<dyn VectorIndexExtension>>;
+}
+
+pub trait ScalarIndexExtension: IndexExtension {
+    // TODO: implement this trait and wire it in
+}
+
 #[async_trait::async_trait]
-pub trait VectorIndexExtension: Send + Sync {
-    /// TODO: add create_index and optimize_index methods
+pub trait VectorIndexExtension: IndexExtension {
+    async fn create_index(
+        &self,
+        // Can't use Arc<Dataset> here
+        // because we need &mut Dataset to call `create_index`
+        // if we wrap into an Arc, the mutable reference is lost
+        dataset: &Dataset,
```

Comment:
Should we just use `&Dataset` in `load_index` as well?

Reply:
we use `Arc<Dataset>` in `open_vector_index_v2`. So it has to be arc for now. I'll take a look and see if we can change it when i do the vector interface upgrade